doc/design/selected_rows.md

```diff
@@ -0,0 +1,61 @@
+# Design Doc: Selected Rows
+
+`SelectedRows` is a kind of sparse tensor data type, which is designed to support `embedding` operators. The gradient of embedding table is a sparse tensor. Only a few rows are non-zero values in that tensor. It is straightforward to represent the sparse tensor by the following sparse tensor data structure:
+
+```cpp
+class SelectedRows {
+ private:
+  vector<int> rows_;
+  Tensor value_;
+  int height_;
+};
+```
+
+The field `height_` shows the first dimension of `SelectedRows`. The `rows` are the indices of which rows of `SelectedRows` are non-zeros. The `value_` field is an N-dim tensor and shape is `[rows.size() /* NUM_ROWS */, ...]`, which supplies values for each row. The dimension of `SelectedRows` satisfies `[height_] + value_.shape[1:]`.
+
+For example, given `height_=100`, `rows_ = [73, 84]`, the `value_ = [[1.0, 2.0], [3.0, 4.0]]` specifies that it is a `100*2` matrix, the 73rd row of that sparse tensor is `[1.0, 2.0]`, the 84th row of that sparse tensor is `[3.0, 4.0]`.
+
+
+## SelectedRows in Protobuf
+
+`SelectedRows` is a kind of `Variable`. `VarDesc` in protobuf should describe the `SelectedRows` information. Only the tensor dimension of a `SelectedRows` will be described in compile-time since the `rows_` and `value_` are related to training data. The `VarDesc` will unify `Dimension` field since `Dimension` is a attribute of both `LoDTensor` and `SelectedRows`.
+
+```proto
+message TensorDesc {
+  required DataType data_type = 1;
+  repeated int64 dims = 2; // [UNK, 640, 480] is saved as [-1, 640, 480]
+}
+
+message VarDesc {
+  required string name = 1;
+  enum VarType { 
+    LoDTensor = 0;
+    SelectedRows = 1;
```

Comment:
Maybe we can also name it `xxTensor` just like `LoDTensor`, for example `SelectedRowsTensor`, because `SelectedRows ` is so different from `Tensor`, hard to make a relationship between them.